worker/discoverspaces/discoverspaces.go

```diff
@@ -140,30 +142,41 @@ func (dw *discoverspacesWorker) handleSubnets() error {
 		logger.Debugf("not a networking environ")
 		return nil
 	}
-	if supported, err := environ.SupportsSpaceDiscovery(); err != nil {
+	canDiscoverSpaces, err := environ.SupportsSpaceDiscovery()
+	if err != nil {
 		return errors.Trace(err)
-	} else if !supported {
-		logger.Debugf("environ does not support space discovery")
-		return nil
 	}
-	providerSpaces, err := environ.Spaces()
+	if canDiscoverSpaces {
+		return errors.Trace(dw.discoverSpaces(environ))
+	}
+	logger.Debugf("environ does not support space discovery")
+	return errors.Trace(dw.discoverSubnetsOnly(environ))
+}
+
+func (dw *discoverspacesWorker) discoverSubnetsOnly(environ environs.NetworkingEnviron) error {
+	stateSubnetIds, err := dw.getStateSubnets()
 	if err != nil {
 		return errors.Trace(err)
 	}
-
-	facade := dw.config.Facade
-	listSpacesResult, err := facade.ListSpaces()
+	subnets, err := environ.Subnets(instance.UnknownId, nil)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	stateSubnets, err := facade.ListSubnets(params.SubnetsFilters{})
+	var addSubnetsArgs params.AddSubnetsParams
+	collectMissingSubnets(&addSubnetsArgs, "", stateSubnetIds, subnets)
```

Comment:
This logic really needs to be server side - it is not transactional as is.
But it seems the existing code suffers from the same problem.....